runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -174,16 +179,19 @@ function LanguageTree:parse()
         vim.list_extend(changes, tree_changes)
```

Comment:
Can't we remove `_regions` and multiple trees per language all together? My understanding was that upstream tree-sitter doesn't even support this concept. It would simplify the implementation.

Reply:
Upstream tree-sitter does not define anything about multiple trees or multiple languages, at all. It is up to us to decide what high level concepts we want here. Forcing multiple regions of the same language to be in the the same parse tree might simplify some stuff but it would imply a parse error in region A will affect independent region B even if they are far apart, just because they are in the same language (think multiple unrelated `vim.cmd` embeddings of vim script inside lua)

Reply:
Just a point of terminology: "upstream" here can mean two different things:

1. the tree-sitter C library (which we use)
2. the tree-sitter Rust reference implementation (which Helix uses but we obviously don't)

I suspect that @theHamsta is referring to 2. while @bfredl is referring to 1. It would help if we made that distinction clearly (e.g., by talking about "upstream library" and "upstream implementation", respectively) to avoid misunderstandings.

Reply:
Well 1 is _our_ upstream here, and is what limits what we can do (without going into any upstream code). Helix does not use the highlighter "reference implementation" as it does not support incremental operation to start with (at least not at the time) so helix implemented they own "layers" just like our LanguageTree to not be limited by the "reference" neither by performance nor by features. I don't we should limit ourselves either.

Reply:
> Well 1 is our upstream here, and is what limits what we can do (without going into any upstream code).

Yes, that is clear. All I'm doing is pointing out that there's danger of miscommunication here; I'm not interested in finger-pointing.

>  Helix does not use the highlighter "reference implementation"

At the very least, it has verbatim copied parts of it, and from what I've seen of its injection queries behaves as similarly to it as I can distinguish.

Reply:
Fair, I just think it would be much more fruitful to directly coordinate with helix (and other _sibling_ editor consumers of upstream tree-sitter) regarding the usecase for editing than some rules that upstream tree-sitter hasn't even set up to define (I know you pretty much agree, this is also for other readers :)

tree-sitter-highlight still takes in a dense `u8[]` and returns a iterator for the entire document. It was and still is implemented to support static highlighting in the context of the github web API, I don't think it was ever meant as a complete reference for highlight in the context of an editor. (you are of course correct it was a useful starting point in a rust codebase)

Reply:
This is exactly what I am doing right now :)
(Coordinating with Helix) 

It just helps me to understand where they are coming from, which is "pure upstream". In particular, I want to know where differences are intentional and need to be preserved vs. what is just "frontend" and could be changed without loss on our end.

Reply:
But I believe tree-sitter has added stuff since our initial implementation a few years ago?